File: go_backend/internal/schemas/video_data.go
package schemas

type VideoData struct {

    //
    Hash           string `json:"hash"`
    DateAdded      string `json:"date_added"`      // when first scanned into DB; set once
    DateDownloaded string `json:"date_downloaded"` // file ModTime at first scan; set once
    Path           string `json:"path"`
    Filename       string `json:"filename"`

    // video attributes
    Duration        string  `json:"duration"`
    DurationSeconds float64 `json:"duration_seconds"`
    FilesizeMB      float64 `json:"filesize_mb"`
    FPS             float64 `json:"fps"`
    Height          int     `json:"height"`
    Width           int     `json:"width"`
    Resolution      string  `json:"resolution"` // "1920x1080"
    AspectRatio     string  `json:"aspect_ratio"`
    Bitrate         int     `json:"bitrate"`
    IsVFR           bool    `json:"is_vfr"`
    VideoCodec      string  `json:"video_codec"`
    AudioCodec      string  `json:"audio_codec"`
    PixFmt          string  `json:"pix_fmt"`
    ColorTransfer   string  `json:"color_transfer"`

    SourceID string `json:"source_id"`

    // collection
    Collection   string `json:"collection"`
    ParentDir    string `json:"parent_dir"`
    PathRelative string `json:"path_relative"`

    // scene attributes
    Title       string `json:"title"`
    SceneTitle  string `json:"scene_title"`
    SceneNumber int    `json:"scene_number"`
    MovieTitle  string `json:"movie_title"`
    MovieSeries string `json:"movie_series"`

    Studio       string `json:"studio"`
    Line         string `json:"line"`
    DateReleased string `json:"date_released"`
    Description  string `json:"description"`
    DVDCode      string `json:"dvd_code"`

    Actors          []string `json:"actors"`
    PrimaryActors   []string `json:"primary_actors"`
    SecondaryActors []string `json:"secondary_actors"`

    // meta
    Tags             []string `json:"tags"`
    TagsFromFilename []string `json:"tags_from_filename"`
    TagsFromPath     []string `json:"tags_from_path"`
    TagsFromJSON     []string `json:"tags_from_json"`
    Genres           []string `json:"genres"`

    // crawled interactions
    Views int `json:"views"`
    Likes int `json:"likes"`

    Metadata map[string]any `json:"metadata"`
}

File: py/lib/schemas/video_data.py
from typing import Optional
from dataclasses import dataclass, field, asdict


@dataclass
class VideoData:
    hash: str
    date_added: str
    path: str
    filename: str

    duration: str
    filesize_mb: float
    fps: float
    height: int
    bitrate: int
    duration_seconds: float

    width: int = 0
    resolution: str = ""
    aspect_ratio: str = ""
    is_vfr: bool = False
    video_codec: str = ""
    audio_codec: str = ""
    pix_fmt: str = ""
    color_transfer: str = ""

    source_id:      Optional[str] = None

    collection:     Optional[str] = None
    parent_dir:     Optional[str] = None
    path_relative:  Optional[str] = None

    title:              Optional[str] = None
    scene_title:        Optional[str] = None
    scene_number:       Optional[int] = None
    movie_title:        Optional[str] = None
    movie_series:       Optional[str] = None

    studio:             Optional[str] = None
    line:               Optional[str] = None
    date_released:      Optional[str] = None
    description:        str = ""
    dvd_code:           Optional[str] = None

    actors:                 list[str] = field(default_factory=list)
    primary_actors:         list[str] = field(default_factory=list)
    secondary_actors:       list[str] = field(default_factory=list)

    tags:               list[str] = field(default_factory=list)
    tags_from_filename: list[str] = field(default_factory=list)
    tags_from_path:     list[str] = field(default_factory=list)
    tags_from_json:     list[str] = field(default_factory=list)
    genres:             list[str] = field(default_factory=list)
    metadata:                dict = field(default_factory=dict)

    views:  Optional[int] = None
    rating: Optional[float] = None
    likes:  Optional[int] = None

    is_linked: bool = True

    def to_dict(self):
        return asdict(self)

    @classmethod
    def from_dict(cls, data, strict=True):
        if strict:
            return cls(**data)
        else:
            valid_keys = {field.name for field in cls.__dataclass_fields__.values()}
            filtered_data = {k: v for k, v in data.items() if k in valid_keys}
            return cls(**filtered_data)
